### PR TITLE
Graphic Helpers Update

### PR DESCRIPTION
**Pull request by dnenov**
### Purpose

Add new preference setting to scale helpers in a graph (Grids, planes, coordinate systems etc.) under Preferences à Visual Settings à Display that ranges from 1 (current) to 10000 to unlock metric users.

There is a[ Dynamo Revit counterpart to this PR](https://github.com/DynamoDS/DynamoRevit/pull/2915), which allows Dynamo to pull current document units (on starting Dynamo only) when using Dynamo inside Revit.

#### Grid scale factor
grid scale 1

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [x] This PR contains no files larger than 50 MB 

### Release Notes

- multiplies the grid cell to scale (up)
- added `CustomColorsPopupTitle` resource inside the corresponding Resources.resx previously missed inside this [PR](https://github.com/DynamoDS/Dynamo/pull/13824)

### Reviewers

@reddyashish 
@QilongTang 

### FYIs

@mmisol

### CODE REVIEW

File: src/DynamoCoreWpf/Views/Menu/PreferencesView.xaml.cs [hunk 1]
```diff
@@ -626,5 +629,32 @@ private void UpdateZoomScaleValueLabel(Slider slider, Label label)
...
+
+        private void UpdateGraphicScaleValueLabel(Slider slider, Label label, bool initialRun = false)
+        {
+            double value = slider.Value;
+            if (label != null)
+            {
+                //The margin value for the label goes from - 480 to 310, resulting in 790 pixels from the starting point to the end.
+                //The value is decreased to 480 because the margin begins at - 480
+                //This is the relation between the margin in pixels and the value
+                double marginValue = (790 * value / 10000) - 480;
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Menu/PreferencesView.xaml.cs:647]: I think any kind of pixel logic is going to need robust testing in various DPI settings - theres probably a way to factor in wpf scale value here.

**Comment by QilongTang** [comment]
@reddyashish @Amoursol Is this intended for 2.18 release?

**Comment by Amoursol** [comment]
@QilongTang no, this can be 2.19.

File: src/DynamoCoreWpf/Views/Menu/PreferencesView.xaml.cs [hunk 2]
```diff
@@ -16,6 +16,7 @@
 using Dynamo.Logging;
 using Dynamo.UI;
 using Dynamo.ViewModels;
+using static Dynamo.UI.Controls.DynamoToolTip;
```
Comment (reddyashish) [inline src/DynamoCoreWpf/Views/Menu/PreferencesView.xaml.cs:19]: No changes to this file, can remove the static usings.

File: src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs [hunk 3]
```diff
@@ -883,6 +975,19 @@ public bool HideAutocompleteMethodOptions
...
+        /// <summary>
+        /// Returns if the current session is Dynamo Revit
+        /// </summary>
+        public bool IsDynamoRevit
+        {
+            get
+            {
+                // HostAnaltyicsInfo is not set when this is invoked??
+                //return this.dynamoViewModel.Model.HostAnalyticsInfo.HostName.Equals("Dynamo Revit");
+                return this.dynamoViewModel.Model.HostName.Equals("Dynamo Revit");
```
Comment (reddyashish) [inline src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs:987]: Model.HostName will be deprecated. Isn't HostAnaltyicsInfo providing the host name? Is that null here?

Comment (dnenov) [reply src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs:987]: That's exactly right. The HostAnalyticsInfo is null when we need it, however the old Model.HostName has the correct value.

Comment (reddyashish) [reply src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs:987]: Can the HostAnalyticsInfo be null and HostName have the correct host in any case? May be HostAnalyticsInfo is set later in the execution?.

File: src/DynamoCoreWpf/Views/Menu/PreferencesView.xaml.cs (hunk 2 above)
Comment (dnenov) [reply src/DynamoCoreWpf/Views/Menu/PreferencesView.xaml.cs:19]: Done!

File: src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs (hunk 3 above)
Comment (dnenov) [reply src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs]: Updated as per @mjkkirschner comment - now checks for HostAnalyticsInfo first, if that is `null` goes on to check Model.HostName.

File: src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs [hunk 4]
```diff
@@ -883,6 +1020,28 @@ public bool HideAutocompleteMethodOptions
...
+        {
+            get
+            {
+                // HostAnalyticsInfo is not set when this is invoked??
+                //return this.dynamoViewModel.Model.HostAnalyticsInfo.HostName.Equals("Dynamo Revit");
+                var host = this.dynamoViewModel.Model.HostAnalyticsInfo.HostName;
+
+                if (host != null)
+                {
+                    return host.Equals("Dynamo Revit");
```
Comment (reddyashish) [inline src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs:1036]: sorry i missed this but i don't think we want to check for the host "Dynamo Revit" by hardcoding it. We have a method GetKnownHosts on PackageManagerClient which can be used to check for the known hosts. @QilongTang What do you think?

**Comment by reddyashish** [comment]
Thanks for updating this Deyan. LGTM with one last comment.

File: src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs (hunk 4 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs:1036]: 1. I don't think that package manager method has the same list of names as the host names in analytics, and 
2. It seems a bit fishy to call a service just to get the name Revit/DynamoRevit which is being hardcoded somewhere else.
What will happen if the service is slow, times out etc, how will that impact Dynamo's performance for this simple check?

File: src/DynamoCore/Configuration/PreferenceSettings.cs [hunk 5]
```diff
@@ -150,6 +158,11 @@ public bool IsADPAnalyticsReportingApproved
         /// </summary>
         public bool ShowPreviewBubbles { get; set; }
 
+        /// <summary>
+        /// Indicates if Revit units should be used for graphic helpers for Dynamo Revit
+        /// </summary>
+        public bool UseRevitScaleUnits { get; set; }
```
Comment (mjkkirschner) [inline src/DynamoCore/Configuration/PreferenceSettings.cs:164]: I think it's strange that Revit should be referenced directly in Dynamo core?

File: src/DynamoCore/Configuration/Configurations.cs [hunk 6]
```diff
@@ -393,6 +393,38 @@ public class Configurations
             }
         }
 
+        public enum Units
```
Comment (mjkkirschner) [inline src/DynamoCore/Configuration/Configurations.cs:396]: does this need to be public?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 7]
```diff
@@ -2466,11 +2466,26 @@ public void ToggleBackgroundGridVisibility(object parameter)
             }
         }
 
+        /// <summary>
+        /// Updates grapic helpers (grid) inside the background preview VM
+        /// </summary>
+        /// <param name="parameter"></param>
+        public void UpdateGraphicHelpersScale(object parameter)
+        {
+            BackgroundPreviewViewModel.GridScale = PreferenceSettings.GridScaleFactor;
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:2475]: could `BackgroundPreviewViewModel` be null? maybe `?`

File: src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs (hunk 4 above)
Comment (dnenov) [reply src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs:1036]: This should be rather simple to achieve. We just need a way to know if we are running under DynamoRevit. It seems it's creating a lot of discussion. Could you, please, point me to a single good solution for it? The goal is to only display a certain user option if Dynamo is used inside Revit. I believe this is the first time something like this is happening (this has been brought up a few times).

To spin this around, if we cannot know if we are under Revit, we will not be able to use this new functionality.

File: src/DynamoCore/Configuration/PreferenceSettings.cs (hunk 5 above)
Comment (dnenov) [reply src/DynamoCore/Configuration/PreferenceSettings.cs:164]: Hey @mjkkirschner - I am afraid this whole process requires for some form of awareness on the Dynamo side that this is a Revit session. Basically, the task is to make the grid layout behave differently under Revit, that's how I understand it. Maybe this core concept is no good, and then we may have to go back a bit and analyze the approach once again? @Amoursol

File: src/DynamoCore/Configuration/Configurations.cs (hunk 6 above)
Comment (dnenov) [reply src/DynamoCore/Configuration/Configurations.cs]: I made them internal, would that be OK?

File: src/DynamoCore/Configuration/Configurations.cs [hunk 8]
```diff
@@ -393,6 +393,38 @@ public class Configurations
...
+            Feet,
+            Inches,
+            Miles
+        }
+
+        /// <summary>
+        /// Supported Revit units and their conversion to dynamo units
+        /// Used in graphic helpers scaling (grid, axes)
+        /// </summary>
+        public static Dictionary<Units, double> SupportedUnits
```
[the remark this answers is not in the document]
Comment (dnenov) [reply src/DynamoCore/Configuration/Configurations.cs]: Same as above, internal.

File: src/DynamoCore/Configuration/PreferenceSettings.cs [hunk 9]
```diff
@@ -240,6 +253,42 @@ public bool IsBackgroundGridVisible
...
+                gridScaleFactor = value;
+
+                RaisePropertyChanged(nameof(GridScaleFactor));
+            }
+        }
+
+        /// <summary>
+        /// The current Revit document units. Will be updated the first time Dynamo is started
+        /// </summary>
+        public Configurations.Units CurrentRevitUnits
```
[the remark this answers is not in the document]
Comment (dnenov) [reply src/DynamoCore/Configuration/PreferenceSettings.cs]: Same as above, internal.

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 7 above)
Comment (dnenov) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:2477]: Yup, got it! Added the check.

File: src/DynamoCore/Configuration/PreferenceSettings.cs (hunk 5 above)
Comment (Amoursol) [reply src/DynamoCore/Configuration/PreferenceSettings.cs:164]: @mjkkirschner agree that having a Revit reference in Core isn't ideal. Is there a way to inject this functionality in from D4R instead?

File: src/DynamoCore/Configuration/PreferenceSettings.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Amoursol) [reply src/DynamoCore/Configuration/PreferenceSettings.cs]: Same comment as https://github.com/DynamoDS/Dynamo/pull/13895#discussion_r1197898057

File: src/DynamoCore/Configuration/PreferenceSettings.cs (hunk 5 above)
Comment (dnenov) [reply src/DynamoCore/Configuration/PreferenceSettings.cs]: Hi @mjkkirschner - I replaced 'Revit' to 'Host' for a platform-agnostic approach, I hope that addresses the concern around semantics, and sorry if I didn't understand correctly where the issue was coming from. I hope I got it right this time around!

File: src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs [hunk 10]
```diff
@@ -791,10 +791,6 @@ private Configurations.Units GetTransformedRevitUnits(Configurations.Units revit
             {
                 return Configurations.Units.Millimeters;
             }
-            else if (revitUnits == Configurations.Units.Inches)
```
Comment (reddyashish) [inline src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs]: Is this not needed?

Comment (dnenov) [reply src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs]: Actually, it turns out that it is not needed. For Imperial units, the direct translation of scale works great. For Metric units, an inverse translation had to take place for things to look 'normal' against the grid. I, then, removed the unnecessary translation for Inches.

**Comment by reddyashish** [comment]
CustomColorPicker_PrePopulateDefaultColors and TestImportCopySettings are failing in the latest job. Please check them once.

**Comment by reddyashish** [comment]
Running the job again now to confirm the failing tests. @mjkkirschner Let us know if you think of any changes needed.

**Pull request merged**